src/toxcore/dht_new/dht_node.rs

```diff
@@ -0,0 +1,1295 @@
+                        self.handle_packet_sendn(packet);
+                        Ok(None)
+                    },
+                    // TODO: handle other kinds of packets
+                    p => {
+                        debug!("Received unhandled packet kind: {:?}", p);
+                        Err(Error::new(ErrorKind::Other,format!("Received unhandled packet kind: {:?}", p)))
+                    },
+                }
+            },
+            &DhtBase::DhtRequest(ref dp) => {
+                let payload = dp.get_payload(&self.sk)
+                    .map(|p| p)
+                    .map_err(|e| {
+                        error!("deserialize DhtRequest payload fail {:?}", e);
+                        return e;
+                    });
+                if let Ok(Some(DhtRequestPayload::NatPingRequest(_))) = payload {
+                    debug!("Received ping request");
+                    self.create_nat_ping_resp(packet, sender_pk)
+                }
+                else {
+                    debug!("Received unhandled packet kind: {:?}", dp);
+                    Err(Error::new(ErrorKind::Other,format!("Received unhandled packet kind: {:?}", dp)))
+                }
+            },
+        }
+    }
+
+    /**
+    Handle [`DhtBase`] that claims to contain [`SendNodes`] packet.
+
+    Packet is dropped if:
+
+    - it doesn't contain [`SendNodes`]
+    - it's not a response to a [`GetNodes`] request (invalid ID)
+
+    [`DhtBase`]: ../dht/enum.DhtBase.html
+    [`DhtPacket`]: ../dht/struct.DhtPacket.html
+    [`GetNodes`]: ../dht/struct.GetNodes.html
+    [`SendNodes`]: ../dht/struct.SendNodes.html
+    */
+    fn handle_packet_sendn(&mut self, packet: &DhtBase) {
+        if let &DhtBase::DhtPacket(ref dp) = packet {
+            let rlt = dp.get_payload(&self.sk)
+            .and_then(|psn| {
+                if let Some(DhtPacketPayload::SendNodes(sn)) = psn {
+                    if self.getn_timeout.remove(sn.id) {
+                        debug!("Received SendN is a valid response");
+                        // received SendNodes packet is a response to our request
+                        trace!("Adding nodes from SendNodes to DhtNode's Kbucket");
+                        for node in &sn.nodes {
+                            self.kbucket.try_add(node);
+                        }
+                    }
+                    Ok(())
+                } else { Ok(())}
+            })
+            .map_err(|_| {
+                error!("Wrong DhtPacket; should have contained SendNodes");
+            });
+            rlt.unwrap_or(())
+        }
+        else {
+            ;
+        }
+    }
+
+    /**
+    Remove nodes that have crossed `secs` timeout threshold.
+    */
+    // TODO: test
+    // TODO: add fn for ping/getn req timeouts with hardcoded consts?
+    pub fn remove_timed_out(&mut self, secs: u64) {
+        for pk in self.getn_timeout.get_timed_out(secs) {
+            debug!("Removing timed out node");
```

Comment:
I think pk should be added to the message, in other case it will be useless to have many same records in a row.

Reply:
We should remove that code from Zetok. It is malformed.